pallets/payment-streams/src/utils.rs

```diff
@@ -547,177 +550,253 @@ where
-                        .is_some(),
-                    ArithmeticError::Overflow
-                );
-
-                // Charge the payment stream from the user's balance
-                T::NativeBalance::transfer(
-                    user_account,
-                    &provider_payment_account,
-                    amount_to_charge,
-                    Preservation::Preserve,
-                )?;
-
-                // Set the last charged price index to be the price index of the last chargeable block
-                DynamicRatePaymentStreams::<T>::mutate(
-                    provider_id,
-                    user_account,
-                    |payment_stream| {
-                        let payment_stream = expect_or_err!(
-                            payment_stream,
-                            "Payment stream should exist if it was found before.",
-                            Error::<T>::PaymentStreamNotFound
-                        );
-                        payment_stream.price_index_when_last_charged =
-                            payment_stream.price_index_at_last_chargeable_block;
-                        Ok::<(), DispatchError>(())
-                    },
-                )?;
-
-                // Update the total amount charged:
-                total_amount_charged = total_amount_charged
-                    .checked_add(&amount_to_charge)
-                    .ok_or(Error::<T>::ChargeOverflow)?;
+                        .ok_or(Error::<T>::ChargeOverflow)?;
+                }
             }
         }
 
         Ok(total_amount_charged)
     }
 
+    /// This function gets the Providers that submitted a valid proof in the last tick using the `ReadProofSubmittersInterface`,
+    /// and updates the last chargeable block and last chargeable price index of those Providers.
+    pub fn do_update_last_chargeable_info(
+        n: BlockNumberFor<T>,
+        weight: &mut frame_support::weights::WeightMeter,
+    ) {
+        // Get last block's number
+        let n = n.saturating_sub(One::one());
+        // Get the Providers that submitted a valid proof in the last tick, if there are any
+        let proof_submitters =
+            <T::ProvidersProofSubmitters as ReadProofSubmittersInterface>::get_proof_submitters_for_block(&n);
+        weight.consume(T::DbWeight::get().reads(1));
+
+        // If there are any proof submitters in the last tick
+        if let Some(proof_submitters) = proof_submitters {
+            // See if we have enough weight limit to process all Providers
+            let weight_per_provider = T::DbWeight::get().reads_writes(1, 1);
+            let weight_limit = weight.remaining();
+            let max_providers = weight_limit
+                .checked_div_per_component(&weight_per_provider)
+                .unwrap_or(0);
+            // If we do have enough weight, update all Providers
+            if proof_submitters.len().try_into().unwrap_or(u64::MAX) <= max_providers {
```

Comment:
This check is not really needed here, as this executes in `on_poll`. Surely there would be enough weight. But more importantly, this HAS to execute here, otherwise providers would have missed their chance to charge for a submitted proof.